rfc: enterprise registration

danhhz opened the pull request:
CockroachDB will include "enterprise features", which (after an
evaluation period) must be purchased to be used by commercial entities.
This document describes the technical mechanisms involved.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14114)
<!-- Reviewable:end -->

danhhz committed:
rfc: enterprise registration

CockroachDB will include "enterprise features", which (after an
evaluation period) must be purchased to be used by commercial entities.
This document describes the technical mechanisms involved.

knz commented on docs/RFCS/enterprise.md:
```diff
@@ -0,0 +1,61 @@
+
+A new system table, `licenses`, contains each license and the timestamp of when
+it was added. It is located in the system config keyspace and so is gossiped to
+every node. When a SQL statement using an enterprise feature is executed, the
```
This needs to be cached. 
A read-only cache with no expiration is sufficient as we can probably assume licenses are not modified in-place.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[vendor, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-Kf8Tm3t5Dd7hb9k4PQ4:-Kf8Tm3t5Dd7hb9k4PQ5:bpx8rdv) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03fee582c371d1c21c3f197b1bfbfd5e01389c0/vendor#L1)):*
> ```
> Subproject commit 9719e4147dbab27024288a89a5ccb883088ff9c3
> ```

Remove `vendor` from the commit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-:-Kf8UtFATRqtFLGnlKss:bnfp4nl)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[vendor, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-Kf8Tm3t5Dd7hb9k4PQ4:-Kf8YQSFr_JgBPPTiz-o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03fee582c371d1c21c3f197b1bfbfd5e01389c0/vendor#L1)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Remove `vendor` from the commit.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114)*
<!-- Sent from Reviewable.io -->

danhhz commented:
@petermattis @spencerkimball This is the continuation of https://github.com/cockroachlabs/enterprise/pull/52 if you want to take a look

arjunravinarayan commented:
*[docs/RFCS/enterprise.md, line 45 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-KfIBQw_88Zi7U_ruf9c:-KfIBQw_88Zi7U_ruf9d:bnxbkvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/066b0014602823b8ad5690c3ccccb81bb1f9b7ac/docs/RFCS/enterprise.md#L45)):*
> ```Markdown
> 
> `BACKUP` and `RESTORE` are a natural fit for this model, but `PARTITION` is more
> difficult as a table may be created using an evaluation license and used after
> ```

This isn't a drawback of this model as much as a drawback of any system with expirable evaluation licenses. Even the CLI alternative would face this same issue.

The only real solution is to be graceful in implementing how any specific enterprise feature deals with an expired license (example of dealing with this gracefully, perhaps *too* gracefully: a user can still restore, with some 'you are violation' nags, but you can't create new backups). Partition is indeed a harder case. One option is to prevent changes to the partitioning config, but still obey it (you certainly don't want to stop partitioning at all, you don't want EU data to leave the EU just because the license was accidentally left to expire for a day or two). It has to be a case-by-case decision for each enterprise feature. We probably want to always nag though. What's the most graceful way to nag a user on an operation that's compatible with "the SQL way"?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/enterprise.md, line 45 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-KfIBQw_88Zi7U_ruf9c:-KfOsdxQXgCjIjNXkNZ_:bqsex0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/066b0014602823b8ad5690c3ccccb81bb1f9b7ac/docs/RFCS/enterprise.md#L45)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

This isn't a drawback of this model as much as a drawback of any system with expirable evaluation licenses. Even the CLI alternative would face this same issue.

The only real solution is to be graceful in implementing how any specific enterprise feature deals with an expired license (example of dealing with this gracefully, perhaps *too* gracefully: a user can still restore, with some 'you are violation' nags, but you can't create new backups). Partition is indeed a harder case. One option is to prevent changes to the partitioning config, but still obey it (you certainly don't want to stop partitioning at all, you don't want EU data to leave the EU just because the license was accidentally left to expire for a day or two). It has to be a case-by-case decision for each enterprise feature. We probably want to always nag though. What's the most graceful way to nag a user on an operation that's compatible with "the SQL way"?
</blockquote></details>

Considering that a user can edit the code and recompile to get around the license (a violation of the license, but possible nonetheless), I would veer towards being permissive here. For example, a valid license is needed to create or edit a partition, but not to use a table with a partition. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-:-KfOt2oAoxIpAd9vulgC:bnfp4nl)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/enterprise.md, line 25 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-KfRkW_XW_Xr4LPbtqig:-KfRkW_XW_Xr4LPbtqih:b6wteot) ([raw file](https://github.com/cockroachdb/cockroach/blob/066b0014602823b8ad5690c3ccccb81bb1f9b7ac/docs/RFCS/enterprise.md#L25)):*
> ```Markdown
> latest entry in this table is selected and both the valid_until date and the
> cluster_id are checked. If the feature is not enabled, a SQL error message is
> returned.
> ```

We could also allow the feature to be used but report to us as well as on the admin UI that the license has expired.

---

*[docs/RFCS/enterprise.md, line 30 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-KfRkwx7H1cXiwUSlD33:-KfRkwx80kPvVrMJk5yv:bsb1w0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/066b0014602823b8ad5690c3ccccb81bb1f9b7ac/docs/RFCS/enterprise.md#L30)):*
> ```Markdown
> CREATE TABLE system.licenses (
>     applied TIMESTAMP PRIMARY KEY,
>     descriptor BYTES
> ```

Can you describe this descriptor as a protobuf somewhere?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/enterprise.md, line 25 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-KfRkW_XW_Xr4LPbtqig:-KfRoH0YPJYqahYcyAd1:bn663v9) ([raw file](https://github.com/cockroachdb/cockroach/blob/066b0014602823b8ad5690c3ccccb81bb1f9b7ac/docs/RFCS/enterprise.md#L25)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

We could also allow the feature to be used but report to us as well as on the admin UI that the license has expired.
</blockquote></details>

I've used other DBs for literally _years_ without ever opening their WebUI, so I don't think we can rely on the UI as the sole means of communication.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/enterprise.md, line 20 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-Kfl5ijWxT6LYpYm4Swu:-Kfl5ijWxT6LYpYm4Swv:bwhll2g) ([raw file](https://github.com/cockroachdb/cockroach/blob/066b0014602823b8ad5690c3ccccb81bb1f9b7ac/docs/RFCS/enterprise.md#L20)):*
> ```Markdown
> serialized and base64 encoded.
> 
> A new system table, `licenses`, contains each license and the timestamp of when
> ```

If/when https://github.com/cockroachdb/cockroach/pull/14230 happens, I think we could just use that to store lic, rather than introducing a whole new table -- history is nice, but we'll want history on settings anyway, so I think there too this can just piggy-back.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/enterprise.md, line 22 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-Kf81oPjwOBgDlgXfeiq-r1-22:-Kfr8DEfTMmQvNgEz5r9:b-dr8dpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03fee582c371d1c21c3f197b1bfbfd5e01389c0/docs/RFCS/enterprise.md#L22)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This needs to be cached. 
A read-only cache with no expiration is sufficient as we can probably assume licenses are not modified in-place.
</blockquote></details>

This is what I meant, sounds like I wasn't explicit enough. Better?

---

*[docs/RFCS/enterprise.md, line 20 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-Kfl5ijWxT6LYpYm4Swu:-Kfr86RhZZw51nghbIMG:bihlpzx) ([raw file](https://github.com/cockroachdb/cockroach/blob/066b0014602823b8ad5690c3ccccb81bb1f9b7ac/docs/RFCS/enterprise.md#L20)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

If/when https://github.com/cockroachdb/cockroach/pull/14230 happens, I think we could just use that to store lic, rather than introducing a whole new table -- history is nice, but we'll want history on settings anyway, so I think there too this can just piggy-back.
</blockquote></details>

I like this a lot. Making a table in the config space just for licensing bothered me a bit.

---

*[docs/RFCS/enterprise.md, line 25 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-KfRkW_XW_Xr4LPbtqig:-Kfr9LymSYQN2HxzZN1w:b4ydpzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/066b0014602823b8ad5690c3ccccb81bb1f9b7ac/docs/RFCS/enterprise.md#L25)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I've used other DBs for literally _years_ without ever opening their WebUI, so I don't think we can rely on the UI as the sole means of communication.
</blockquote></details>

Put this in Unresolved questions

---

*[docs/RFCS/enterprise.md, line 30 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-KfRkwx7H1cXiwUSlD33:-Kfr80hoht_tY0kkX_Ls:b-q8nubv) ([raw file](https://github.com/cockroachdb/cockroach/blob/066b0014602823b8ad5690c3ccccb81bb1f9b7ac/docs/RFCS/enterprise.md#L30)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Can you describe this descriptor as a protobuf somewhere?
</blockquote></details>

Yup. Here's a strawman

---

*[docs/RFCS/enterprise.md, line 45 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-KfIBQw_88Zi7U_ruf9c:-Kfr7u5oPcfxBGo7s5mk:b4fh7rr) ([raw file](https://github.com/cockroachdb/cockroach/blob/066b0014602823b8ad5690c3ccccb81bb1f9b7ac/docs/RFCS/enterprise.md#L45)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Considering that a user can edit the code and recompile to get around the license (a violation of the license, but possible nonetheless), I would veer towards being permissive here. For example, a valid license is needed to create or edit a partition, but not to use a table with a partition. 
</blockquote></details>

Added info about partition.

@arjunravinarayan, that's a good point about the nagging. I added it to an Unresolved questions section

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114)*
<!-- Sent from Reviewable.io -->

shivshankardayal commented:
So community users cannot use these features even though they make nothing from using this.

danhhz commented:
@shivshankardayal we're planning to offer free licenses for non-commercial use

shivshankardayal commented:
@danhhz Can I make little bit of money from ads for server costs? Non-profit site or for charity.

danhhz commented:
@shivshankardayal Sorry this took me a bit, I had to look it up. According to our [pricing page](https://www.cockroachlabs.com/pricing/), "CockroachDB Enterprise is free for non-commercial use (e.g., for personal, nonprofit and educational use)."

knz commented:
Reviewed 1 of 2 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/enterprise.md, line 22 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-Kf81oPjwOBgDlgXfeiq-r1-22:-KfrfvGiWosOSogLag7T:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03fee582c371d1c21c3f197b1bfbfd5e01389c0/docs/RFCS/enterprise.md#L22)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

This is what I meant, sounds like I wasn't explicit enough. Better?
</blockquote></details>

:+1:

---

*[docs/RFCS/enterprise.md, line 25 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-KfRkW_XW_Xr4LPbtqig:-Kfrfy56YeLtwu5-_cCs:bnjv4ej) ([raw file](https://github.com/cockroachdb/cockroach/blob/066b0014602823b8ad5690c3ccccb81bb1f9b7ac/docs/RFCS/enterprise.md#L25)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Put this in Unresolved questions
</blockquote></details>

I would suggest that we keep the license files in the on-db table, and make admins able to SELECT through this table to view their past licenses.

---

*[docs/RFCS/enterprise.md, line 71 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-KfrgAt_U0HgVkmVP8yK:-KfrgAtaVtXQj2bl7zWY:bilnarz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1d100eccf6e72a66504a373fe98694f19ceb324/docs/RFCS/enterprise.md#L71)):*
> ```Markdown
> When a user's license is expired (or near expired), they need to be notified.
> Assuming that not every customer will regularly use the admin ui, how should we
> do these sorts of notifications?
> ```

During node start-up; the same place we report candidate updates for the binary.
Possibly also using a periodic process, in the logs. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114)*
<!-- Sent from Reviewable.io -->

benesch commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/enterprise.md, line 71 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-KfrgAt_U0HgVkmVP8yK:-KfxMciw2BjJV7qaXj5g:b-2cw2ym) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1d100eccf6e72a66504a373fe98694f19ceb324/docs/RFCS/enterprise.md#L71)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

During node start-up; the same place we report candidate updates for the binary.
Possibly also using a periodic process, in the logs. 
</blockquote></details>

We probably also want to send an email to the cluster administrator (address provided when they register for the license). That's entirely out of scope, of course.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-:-KfxNGRqCN6TlgMShFaY:bnfp4nl)*
<!-- Sent from Reviewable.io -->

danhhz commented:
@bdarnell are you okay with switching to storing this in the settings table from #14230 instead of the originally proposed new system table that's just for registrations?

---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[docs/RFCS/enterprise.md, line 25 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-KfRkW_XW_Xr4LPbtqig:-Kg0tEXJtHDNY1NG2tgX:b9x5o1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/066b0014602823b8ad5690c3ccccb81bb1f9b7ac/docs/RFCS/enterprise.md#L25)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I would suggest that we keep the license files in the on-db table, and make admins able to SELECT through this table to view their past licenses.
</blockquote></details>

Yup. This is what I meant by "Once viewing historical settings is supported, it can be used to show historical licensing information". I'm intentionally keeping it vague because that rfc hasn't specified a syntax yet for viewing historical values, though it'll almost certainly be a SELECT over a system table in the non-config id space.

---

*[docs/RFCS/enterprise.md, line 71 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-KfrgAt_U0HgVkmVP8yK:-Kg0toQDEdbPjlrQPEUz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1d100eccf6e72a66504a373fe98694f19ceb324/docs/RFCS/enterprise.md#L71)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

We probably also want to send an email to the cluster administrator (address provided when they register for the license). That's entirely out of scope, of course.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14114#-:-Kg0u0W__oj0U6eDnM60:b-735kkj)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
I don't feel too strongly about it, but I'd still prefer putting this in a separate table. I know we're not being very strict about the enforcement here, but it still feels wrong to mix the license data in with (world-readable?) settings and making it trivial for any user of a cluster to get that cluster's license. I think it's worth making a separate table.

danhhz commented:
@bdarnell Yeah, it does seem wrong for a user of a shared cluster to be able to get the cluster license. If we made the settings table root-only, does that address your concern? That seems to me to be a good idea anyway.

Not pushing back, I don't have a strong opinion either way yet. The tradeoff is a possible impedance mismatch between registrations and settings vs using another of a very limited set of system table ids. Even if we make settings root-only, there might be enough of a difference between them and registrations that we make a separate table for that reason

bdarnell commented:
My intuition is that the settings table should be world-readable (or maybe the table is root-only and (most of) the individual settings are exposed to all users via `SHOW`?). I'd lean towards a separate table because licenses are different enough from other settings that we're likely to want to treat them differently in the future. However, I had forgotten how severely limited our gossiped IDs are (only 10; we're using 5 today and the settings table will be 6). So we should probably use the settings table for now and create a separate table when/if we need it.

danhhz commented:
wfm, added a note that we may eventually make a new system table for this
